components/raftstore/src/store/fsm/store.rs

```diff
@@ -1534,22 +1535,38 @@ impl<EK: KvEngine, ER: RaftEngine> RaftBatchSystem<EK, ER> {
             io_reschedule_concurrent_count: Arc::new(AtomicUsize::new(0)),
         };
         let region_peers = builder.init()?;
-        self.start_system::<T, C>(
-            workers,
-            region_peers,
-            builder,
-            auto_split_controller,
-            concurrency_manager,
-            mgr,
-            pd_client,
-            collector_reg_handle,
-            region_read_progress,
-            health_service,
-        )?;
+        let engine = builder.engines.kv.clone();
+        if engine.support_write_batch_vec() {
+            self.start_system::<T, C, <EK as WriteBatchExt>::WriteBatchVec>(
+                workers,
+                region_peers,
+                builder,
+                auto_split_controller,
+                concurrency_manager,
+                mgr,
+                pd_client,
+                collector_reg_handle,
+                region_read_progress,
+                health_service,
+            )?;
+        } else {
+            self.start_system::<T, C, <EK as WriteBatchExt>::WriteBatch>(
+                workers,
+                region_peers,
+                builder,
+                auto_split_controller,
+                concurrency_manager,
+                mgr,
+                pd_client,
+                collector_reg_handle,
+                region_read_progress,
+                health_service,
+            )?;
+        }
         Ok(())
     }
 
-    fn start_system<T: Transport + 'static, C: PdClient + 'static>(
+    fn start_system<T: Transport + 'static, C: PdClient + 'static, W: WriteBatch<EK> + 'static>(
```

Comment:
I don't like this generic type either. I suggest let `RocksWriteBatchVec` become the new `RocksEngine::WriteBatch`. When engine doesn't support multi_batch, we simple do nothing in `check_switch_batch`, and write will always use `write_opt`.

Reply:
Thanks, I'll refactor this part.